main/src/library/Fixpoint/Compiler.flix

```diff
@@ -0,0 +1,258 @@
+            /// Compile rule into an incremental join loop where delta designates the focused atom.
+            let compile = delta ->
+                Array.foldRight(match (atom, rowVar) -> acc -> match atom {
+                    case BodyAtom(PredSym(pred, _), den, Polarity.Positive, _, _) =>
+                        let relSym = if (rowVar == delta) {
+                            RelSym.Delta(pred, den)
+                        } else {
+                            RelSym.Full(pred, den)
+                        };
+                        RelOp.Search(rowVar, relSym, acc)
+                    case _ => acc
+                }, loopBody, augBody) as & Pure |>
+                RamStmt.Insert;
+            let positiveAtoms = Array.filter(match (atom, _) -> match atom {
+                case BodyAtom(_, _, Polarity.Positive, _, _) => true
+                case _ => false
+            }, augBody) as & Pure;
+            let deltas = Array.fold(Array.map(match (_, rowVar) -> compile(rowVar), positiveAtoms)) as & Pure;
+            let comment = RamStmt.Comment(ToString.toString(rule));
+            List.fold(comment :: deltas :: Nil)
+    }
+
+    def compileHeadTerm(env: Map[VarSym, RamTerm[v]], term: HeadTerm[v]): RamTerm[v] = match term {
+        case HeadTerm.Var(var, _) => unwrap(Map.get(var, env))
+        case HeadTerm.Lit(v, _) => RamTerm.Lit(v)
+        case HeadTerm.App0(f, _) => RamTerm.App0(f)
+        case HeadTerm.App1(f, v, _) =>
+            let t = unwrap(Map.get(v, env));
+            RamTerm.App1(f, t)
+        case HeadTerm.App2(f, v1, v2, _) =>
+            let t1 = unwrap(Map.get(v1, env));
+            let t2 = unwrap(Map.get(v2, env));
+            RamTerm.App2(f, t1, t2)
+        case HeadTerm.App3(f, v1, v2, v3, _) =>
+            let t1 = unwrap(Map.get(v1, env));
+            let t2 = unwrap(Map.get(v2, env));
+            let t3 = unwrap(Map.get(v3, env));
+            RamTerm.App3(f, t1, t2, t3)
+        case HeadTerm.App4(f, v1, v2, v3, v4, _) =>
+            let t1 = unwrap(Map.get(v1, env));
+            let t2 = unwrap(Map.get(v2, env));
+            let t3 = unwrap(Map.get(v3, env));
+            let t4 = unwrap(Map.get(v4, env));
+            RamTerm.App4(f, t1, t2, t3, t4)
+        case HeadTerm.App5(f, v1, v2, v3, v4, v5, _) =>
+            let t1 = unwrap(Map.get(v1, env));
+            let t2 = unwrap(Map.get(v2, env));
+            let t3 = unwrap(Map.get(v3, env));
+            let t4 = unwrap(Map.get(v4, env));
+            let t5 = unwrap(Map.get(v5, env));
+            RamTerm.App5(f, t1, t2, t3, t4, t5)
+    }
+
+    /// Augment body atoms with row variables.
+    def augmentBody(body: Array[BodyPredicate[v]]): Array[(BodyPredicate[v], RowVar)] =
+        Array.mapWithIndex(atom -> i -> match atom {
+            case BodyAtom(predSym, _, Polarity.Positive, _, _) => (atom, RowVar.Named("${predSym}$${i}"))
+            case _ => (atom, RowVar.Named("IfYouSeeThisYouFoundABug"))
```

Comment:
Why is this not a `bug!`?

Reply:
Because it's (supposed to be) unused. Negative atoms and guards are still valid, I just don't use their associated row variable.